### PR TITLE
Fix lifespan tests

### LINKED ISSUE #911: Lifespan tests are broken
### Checklist

- [x] The bug is reproducible against the latest release and/or `master`.
- [x] There are no similar issues or pull requests to fix it yet.

### Describe the bug

Some lifespan tests are passing fine while they should fail

### To reproduce

this one for instance pass while the log suggests it should fail, if you change the assert scope to assert False it still passes !

```
@pytest.mark.parametrize("mode", ("auto", "on"))
def test_lifespan_scope_asgi3app(mode):
    async def asgi3app(scope, receive, send):
        assert scope == {"version": "3.0", "spec_version": "2.0"}

    async def test():
        config = Config(app=asgi3app, lifespan=mode)
        lifespan = LifespanOn(config)

        await lifespan.startup()
        await lifespan.shutdown()

    loop = asyncio.new_event_loop()
    loop.run_until_complete(test())
```

```
/home/lotso/PycharmProjects/uvicorn/venv/bin/python /home/lotso/.local/share/JetBrains/Toolbox/apps/PyCharm-P/ch-0/203.5981.165/plugins/python/helpers/pycharm/_jb_pytest_runner.py --target test_lifespan.py::test_lifespan_scope_asgi3app
Testing started at 2:47 PM ...
Launching pytest with arguments test_lifespan.py::test_lifespan_scope_asgi3app in /home/lotso/PycharmProjects/uvicorn/tests

============================= test session starts ==============================
platform linux -- Python 3.8.5, pytest-6.2.0, py-1.10.0, pluggy-0.13.1 -- /home/lotso/PycharmProjects/uvicorn/venv/bin/python
cachedir: .pytest_cache
rootdir: /home/lotso/PycharmProjects/uvicorn, configfile: setup.cfg
plugins: reraise-1.0.3, asyncio-0.14.0
collecting ... collected 2 items

test_lifespan.py::test_lifespan_scope_asgi3app[auto] 
test_lifespan.py::test_lifespan_scope_asgi3app[on] 

============================== 2 passed in 0.11s ===============================

Process finished with exit code 0
PASSED              [ 50%]INFO:     Waiting for application startup.
INFO:     ASGI 'lifespan' protocol appears unsupported.
INFO:     Application startup complete.
PASSED                [100%]INFO:     Waiting for application startup.
ERROR:    Exception in 'lifespan' protocol
Traceback (most recent call last):
  File "/home/lotso/PycharmProjects/uvicorn/uvicorn/lifespan/on.py", line 55, in main
    await app(scope, self.receive, self.send)
  File "/home/lotso/PycharmProjects/uvicorn/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/home/lotso/PycharmProjects/uvicorn/tests/test_lifespan.py", line 161, in asgi3app
    assert scope == {"version": "3.0", "spec_version": "2.0"}
AssertionError: assert {'asgi': {'spec_version': '2.0', 'version': '3.0'}, 'type': 'lifespan'} == {'spec_version': '2.0', 'version': '3.0'}
ERROR:    Application startup failed. Exiting.
```

### Expected behavior

test should fail

### Environment

- OS / Python / Uvicorn version: just run `uvicorn --version`
- The exact command you're running uvicorn with, all flags you passed included. If you run it with gunicorn please do the same. If there is a reverse-proxy involved and you cannot reproduce without it please give the minimal config of it to reproduce.

### PR DESCRIPTION
**Pull request by euri10**
Fixes #911

it seems to me we have to raise the lifespan exception should it occur and call the lifespan main coroutine in the tests.

### CODE REVIEW

File: tests/test_lifespan.py [hunk 1]
```diff
@@ -158,13 +160,17 @@ async def test():
...
+            "asgi": {"spec_version": "2.0", "version": "3.0"},
+            "type": "lifespan",
+        }
 
     async def test():
         config = Config(app=asgi3app, lifespan=mode)
         lifespan = LifespanOn(config)
 
         await lifespan.startup()
+        await lifespan.main()
```
Comment (florimondmanca) [inline tests/test_lifespan.py:173]: Hmm, this feels weird. `lifespan.startup()` already schedules `lifespain.main()` on the event loop:

https://github.com/encode/uvicorn/blob/a9a46e87d8ba41f08b3a401b7b0cc572333a7193/uvicorn/lifespan/on.py#L29

I think we should instead add assertions:

```suggestion
        assert not lifespan.startup_failed
        assert not lifespan.error_occurred
        assert not lifespan.should_exit
```

There's also a small inconsistency here bc we test against `mode=auto` _and_ `mode=on`, but we always use `LifespanOn`. I think we don't necessarily need to bother with lifespan modes, just test the scope has the correct shape against `mode=on` (that would then properly show any exceptions in the error logs).

File: tests/test_lifespan.py [hunk 2]
```diff
@@ -186,6 +195,7 @@ async def test():
...
```
Comment (florimondmanca) [inline tests/test_lifespan.py:198]:
```suggestion
        await lifespan.startup_event.wait()
```

File: tests/test_lifespan.py (hunk 1 above)
Comment (florimondmanca) [reply tests/test_lifespan.py:173]: Filed #913 to illustrate. :)

Comment (euri10) [reply tests/test_lifespan.py:173]: yes this is way better I missed that error_occured !

**Pull request closed without merging**